### PR TITLE
Adds OW_BERRY_IMMORTAL

### PR DESCRIPTION
**Pull request by pkmnsnfrn**
# Description
Brendan plants an Oran Berry, cycles trees forward five times and it doesn't die

* Adds `OW_BERRY_IMMORTAL`. If enabled, once a Berry tree has grown a berry, the tree will not disappear until picked by the player.

#### `OW_BERRY_IMMORTAL`
Developers must change this value to TRUE in `include/config/overworld.h` to enable.

# Testing
## Clean Branch
You can recreate this branch by applying a patch or pulling the repo. From a clean version of expansion's upcoming, you can either:

### Patch
`wget https://files.catbox.moe/inkkw3.patch -O immortal.patch ; git apply immortal.patch ; rm immortal.patch`

### Repo
`git remote add psf-expansion https://github.com/PokemonSanFran/pokeemerald-expansion/ ; git pull psf-expansion immortalBerries`

## Manual Tests
After replicating the branch, to recreate my testing environment, you can either directly download the debug script, or manually create the changes.

#### `OW_BERRY_IMMORTAL == TRUE`
`wget https://files.catbox.moe/2oij37.h -O include/config/overworld.h`
#### `OW_BERRY_IMMORTAL == FALSE`
`wget https://files.catbox.moe/fv3nbk.h -O include/config/overworld.h`

### Manual Testing
* Modify `include/config/overworld.h` to the desired values
* Compile
* Start new game
* Debug > Utilities > Warp to Map Warp > 0 > 19 > 1
* Pick Oran berry
* Plant Oran Berry
* Debug > Utilities > Berry Functions > Grow map trees (run this five times)

## Verified Scenarios
All videos show the steps described in "Manual Testing"
### TRUE

https://github.com/user-attachments/assets/0e394de0-72e7-416a-99e6-73e29416e75c

### FALSE

https://github.com/user-attachments/assets/58e15eb7-790a-4e25-8c45-1db79b84c4db

# **People who collaborated with me in this PR**
This was originally [written by](<https://www.pokecommunity.com/threads/simple-modifications-directory.416647/page-4#post-10142996>) by @BuffelSaft. Give them all the credit.
# Discord Contact Info
I am `pkmnsnfrn` on Discord.

### CODE REVIEW

File: include/config/overworld.h [hunk 1]
```diff
@@ -33,6 +33,7 @@
 #define OW_BERRY_GROWTH_RATE           GEN_3      // Presets for how long each Berry plant takes to grow.
 #define OW_BERRY_YIELD_RATE            GEN_3      // Presets for how many Berries each plant can yield.
 #define OW_BERRY_DRAIN_RATE            GEN_6_ORAS // If OW_BERRY_MOISTURE is enabled, this setting changes how fast the soil dries out. GEN_4 uses a Berry-dependent drain rate, GEN_6_XY dries out in 24 hours (4 hours with the relevant Mulch) and GEN_6_ORAS dries out in 4 hours. Other values are illegal.
+#define OW_BERRY_IMMORTAL              FALSE // If enabled, once a Berry tree has grown a berry, the tree will not disappear until picked by the player.
```
Comment (Bassoonian) [inline include/config/overworld.h:36]:
```suggestion
#define OW_BERRY_DRAIN_RATE            GEN_6_ORAS // If OW_BERRY_MOISTURE is enabled, this setting changes how fast the soil dries out. GEN_4 uses a Berry-dependent drain rate, GEN_6_XY dries out in 24 hours (4 hours with the relevant Mulch) and GEN_6_ORAS dries out in 4 hours. Other values are illegal.
#define OW_BERRY_IMMORTAL              FALSE     // If enabled, once a Berry tree has grown a Berry, the tree will not disappear until picked by the player.
```

File: src/berry.c [hunk 2]
```diff
@@ -1842,16 +1844,16 @@ static u16 GetMulchAffectedGrowthRate(u16 berryDuration, u8 mulch, u8 stage)
...
 
     for (i = 0; i < BERRY_TREES_COUNT; i++)
     {
         tree = &gSaveBlock1Ptr->berryTrees[i];
 
-        if (tree->berry && tree->stage && !tree->stopGrowth)
+        if (tree->berry && tree->stage && !tree->stopGrowth && (!OW_BERRY_IMMORTAL || tree->stage != BERRY_STAGE_BERRIES))
         {
-            if (minutes >= GetStageDurationByBerryType(tree->berry) * 71)
+            if ((minutes >= GetStageDurationByBerryType(tree->berry) * 71) && !OW_BERRY_IMMORTAL)
```
Comment (Bassoonian) [inline src/berry.c:1856]: Add the check for the constant at the beginning of the condition, this way, the compiler will optimise out this entire block if the config would always result in the condition being false

Comment (pkmnsnfrn) [reply src/berry.c]: https://github.com/rh-hideout/pokeemerald-expansion/pull/5187/commits/08173147c47658a390c91de65f931d46a9588b16

File: include/config/overworld.h [hunk 3]
```diff
@@ -33,6 +33,7 @@
 #define OW_BERRY_GROWTH_RATE           GEN_3      // Presets for how long each Berry plant takes to grow.
 #define OW_BERRY_YIELD_RATE            GEN_3      // Presets for how many Berries each plant can yield.
 #define OW_BERRY_DRAIN_RATE            GEN_6_ORAS // If OW_BERRY_MOISTURE is enabled, this setting changes how fast the soil dries out. GEN_4 uses a Berry-dependent drain rate, GEN_6_XY dries out in 24 hours (4 hours with the relevant Mulch) and GEN_6_ORAS dries out in 4 hours. Other values are illegal.
+#define OW_BERRY_IMMORTAL              FALSE     // If enabled, once a Berry tree has grown a Berry, the tree will not disappear until picked by the player.
```
Comment (Bassoonian) [inline include/config/overworld.h:36]:
```suggestion
#define OW_BERRY_IMMORTAL              FALSE      // If enabled, once a Berry tree has grown a Berry, the tree will not disappear until picked by the player.
```
I missed a space, sorry

**Review by Bassoonian — APPROVED** [review]

**Comment by Bassoonian** [comment]
@pkmnsnfrn can you merge upcoming into this again? #5193 is needed to make this compile again

**Pull request merged**